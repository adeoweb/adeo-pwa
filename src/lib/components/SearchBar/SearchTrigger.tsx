import React, { FunctionComponent } from 'react';

import { useSearchTrigger } from '@magento/peregrine/lib/talons/Header/useSearchTrigger';

type TSearchTriggerProps = {
    onClick: () => void;
};

const SearchTrigger: FunctionComponent<TSearchTriggerProps> = ({ onClick }) => {
    const { handleClick } = useSearchTrigger({
        onClick
    });

    return (
        <button className="search-toggle btn" onClick={handleClick}>
            <i className="icon-magnifier" />
        </button>
    );
};

export default SearchTrigger;
