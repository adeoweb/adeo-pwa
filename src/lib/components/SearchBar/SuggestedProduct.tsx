import React, { FunctionComponent, useCallback } from 'react';
import { useTranslation } from 'react-i18next';

import Price from '@magento/peregrine/lib/Price';

import Image from 'src/lib/components/Image';
import RichText from 'src/lib/components/RichText';
import { Link } from 'src/lib/drivers';
import { TProductSearch } from 'src/lib/types/ProductSearch';
import getItemUrl from 'src/lib/util/getItemUrl';

export type TSuggestedProductProps = {
    onNavigate: () => void;
} & TProductSearch;

const IMAGE_WIDTH = 60;

const SuggestedProduct: FunctionComponent<TSuggestedProductProps> = product => {
    const { t } = useTranslation('product');
    const {
        small_image: smallImage,
        name,
        onNavigate,
        price_range: priceRange,
        short_description: shortDescription
    } = product;

    const handleClick = useCallback(() => {
        if (typeof onNavigate === 'function') {
            onNavigate();
        }
    }, [onNavigate]);

    const imageResource =
        smallImage && smallImage.url ? smallImage.url : undefined;

    const uri = getItemUrl(product);

    let price: React.ReactNode | null = null;
    if (
        priceRange?.maximum_price?.final_price?.currency &&
        priceRange?.maximum_price?.final_price?.value
    ) {
        const { currency, value } = priceRange.maximum_price.final_price;
        price = <Price currencyCode={currency} value={value} />;
    }

    return (
        <Link to={uri} onClick={handleClick}>
            <div className="suggestion-left">
                <Image
                    alt={name || ''}
                    resource={imageResource}
                    width={IMAGE_WIDTH}
                />
            </div>
            <div className="suggestion-right">
                <div className="product-line product-name">{name}</div>
                <div className="product-line product-price">
                    {t('Price')}: {price}
                </div>
                <div className="product-des">
                    <p className="short-des">
                        {shortDescription && shortDescription.html && (
                            <RichText content={shortDescription.html} />
                        )}
                    </p>
                </div>
            </div>
        </Link>
    );
};

export default SuggestedProduct;
