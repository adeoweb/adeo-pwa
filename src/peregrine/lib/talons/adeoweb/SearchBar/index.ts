export { useSearchField } from './useSearchField';
export { useSearchBar } from './useSearchBar';
export { useAutocomplete } from './useAutocomplete';
