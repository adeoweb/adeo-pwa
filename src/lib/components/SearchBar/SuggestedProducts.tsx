import React, { Fragment, FunctionComponent } from 'react';

import { TProductSearch } from 'src/lib/types/ProductSearch';

import SuggestedProduct from './SuggestedProduct';

export type TSuggestedProductsProps = {
    products: TProductSearch[];
    onNavigate: () => void;
    limit?: number;
};

const SuggestedProducts: FunctionComponent<TSuggestedProductsProps> = ({
    limit = 3,
    onNavigate,
    products
}) => {
    const items = products.slice(0, limit).map(product => (
        <div key={product?.id} className="autocomplete-suggestion">
            <SuggestedProduct {...product} onNavigate={onNavigate} />
        </div>
    ));

    return <Fragment>{items}</Fragment>;
};

export default SuggestedProducts;
