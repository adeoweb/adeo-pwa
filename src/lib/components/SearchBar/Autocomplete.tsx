import React, { FunctionComponent } from 'react';
import { useTranslation } from 'react-i18next';

import PRODUCT_SEARCH from '../../queries/productSearch.graphql';
import { useAutocomplete } from 'src/peregrine/lib/talons/adeoweb/SearchBar/useAutocomplete';
import Suggestions from 'src/lib/components/SearchBar/Suggestions';

type TAutocompleteProps = {
    setVisible: (isVisible: boolean) => void;
    visible: boolean;
    searchValue: string;
};

const MESSAGES = new Map()
    .set('ERROR', 'An error occurred while fetching results.')
    .set('LOADING', 'Fetching results...')
    .set('PROMPT', 'Search for a product')
    .set('EMPTY_RESULT', 'No results were found.');

const Autocomplete: FunctionComponent<TAutocompleteProps> = ({
    setVisible,
    visible,
    searchValue
}) => {
    const { t } = useTranslation();
    const { messageType, products } = useAutocomplete({
        query: PRODUCT_SEARCH,
        visible,
        value: searchValue
    });
    const wrapperClass = `autocomplete-suggestions${visible ? ' show' : ''}`;
    const message = t(MESSAGES.get(messageType));

    return (
        <div className={wrapperClass}>
            {message && <div className="autocomplete-title">{message}</div>}
            <Suggestions
                products={products}
                setVisible={setVisible}
                visible={visible}
            />
        </div>
    );
};

export default Autocomplete;
