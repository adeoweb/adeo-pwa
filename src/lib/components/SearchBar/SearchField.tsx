import React, { ChangeEvent, Fragment, FunctionComponent } from 'react';
import { useTranslation } from 'react-i18next';
import * as H from 'history';
import { useSearchField } from 'src/peregrine/lib/talons/adeoweb/SearchBar/useSearchField';

type TSearchFieldProps = {
    onChange: (event: ChangeEvent<HTMLInputElement>) => void;
    onFocus: () => void;
    location: H.Location;
    value: string;
    setFieldValue: (field: string, value: string) => void;
};

const SearchField: FunctionComponent<TSearchFieldProps> = ({
    location,
    onChange,
    onFocus,
    value,
    setFieldValue
}) => {
    const { t } = useTranslation();
    useSearchField({ location, setFieldValue });

    return (
        <Fragment>
            <input
                className="form-control"
                name="search_query"
                onFocus={onFocus}
                onChange={onChange}
                value={value}
                placeholder={t('Search...')}
            />
            <button className="search-button btn">
                <i className="icon-magnifier" />
            </button>
        </Fragment>
    );
};

export default SearchField;
