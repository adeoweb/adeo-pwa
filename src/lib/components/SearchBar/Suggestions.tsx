import React, { FunctionComponent } from 'react';
import SuggestedProducts from './SuggestedProducts';
import { TProductsSearch } from 'src/lib/types/ProductSearch';
import { useSuggestions } from 'src/peregrine/lib/talons/adeoweb/SearchBar/useSuggestions';

type TSuggestionsProps = {
    products?: TProductsSearch;
    setVisible: (isVisible: boolean) => void;
    visible: boolean;
};

const Suggestions: FunctionComponent<TSuggestionsProps> = ({
    products,
    setVisible,
    visible
}) => {
    const items = products && products.items ? products.items : [];

    const { onNavigate, shouldRender } = useSuggestions({
        items,
        setVisible,
        visible
    });

    if (!shouldRender) {
        return null;
    }

    return <SuggestedProducts onNavigate={onNavigate} products={items} />;
};

export default Suggestions;
