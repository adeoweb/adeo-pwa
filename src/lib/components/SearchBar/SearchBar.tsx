import React, { FunctionComponent } from 'react';
import SearchField from './SearchField';
import SearchTrigger from './SearchTrigger';
import { Form } from 'react-bootstrap';
import * as H from 'history';
import Autocomplete from './Autocomplete';
import { useSearchBar } from 'src/peregrine/lib/talons/adeoweb/SearchBar/useSearchBar';

type TSearchBarProps = {
    handleTriggerClick: () => void;
    history: H.History;
    location: H.Location;
    isOpen: boolean;
};

const SearchBar: FunctionComponent<TSearchBarProps> = ({
    handleTriggerClick,
    history,
    isOpen,
    location
}) => {
    const {
        containerRef,
        handleFocus,
        setExpanded,
        expanded,
        form
    } = useSearchBar({
        history
    });
    const { values, handleSubmit, handleChange, setFieldValue } = form;
    const { search_query: searchQuery } = values;
    const wrapperClass = `header-search-wrapper${isOpen ? ' show' : ''}`;

    return (
        <div className="header-search" ref={containerRef}>
            <SearchTrigger onClick={handleTriggerClick} />
            <Form autoComplete="off" onSubmit={handleSubmit}>
                <div className={wrapperClass}>
                    <SearchField
                        location={location}
                        onChange={handleChange}
                        onFocus={handleFocus}
                        value={searchQuery}
                        setFieldValue={setFieldValue}
                    />
                    <Autocomplete
                        setVisible={setExpanded}
                        visible={expanded}
                        searchValue={searchQuery}
                    />
                </div>
            </Form>
        </div>
    );
};

export default SearchBar;
