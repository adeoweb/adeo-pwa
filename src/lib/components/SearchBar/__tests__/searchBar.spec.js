import React from 'react';
import { createTestInstance } from '@magento/peregrine';

import SearchBar from '../SearchBar';

jest.mock('../Autocomplete', () => () => null);
jest.mock('../SearchField', () => () => null);

const handleTriggerClick = jest.fn();

test('renders correctly when closed', () => {
    const tree = createTestInstance(
        <SearchBar
            history={{}}
            isOpen={false}
            location={{}}
            handleTriggerClick={handleTriggerClick}
        />
    );

    expect(tree.toJSON()).toMatchSnapshot();
});

test('renders correctly when open', () => {
    const tree = createTestInstance(
        <SearchBar
            history={{}}
            isOpen={true}
            location={{}}
            handleTriggerClick={handleTriggerClick}
        />
    );

    expect(tree.toJSON()).toMatchSnapshot();
});
