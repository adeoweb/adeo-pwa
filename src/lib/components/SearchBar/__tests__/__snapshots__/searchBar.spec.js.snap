// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when closed 1`] = `
<div
  className="header-search"
>
  <button
    className="search-toggle btn"
    onClick={[Function]}
  >
    <i
      className="icon-magnifier"
    />
  </button>
  <form
    autoComplete="off"
    className=""
    onSubmit={[Function]}
  >
    <div
      className="header-search-wrapper"
    />
  </form>
</div>
`;

exports[`renders correctly when open 1`] = `
<div
  className="header-search"
>
  <button
    className="search-toggle btn"
    onClick={[Function]}
  >
    <i
      className="icon-magnifier"
    />
  </button>
  <form
    autoComplete="off"
    className=""
    onSubmit={[Function]}
  >
    <div
      className="header-search-wrapper show"
    />
  </form>
</div>
`;
